### PR TITLE
QIR unit tests for partial applications

### PR DESCRIPTION
**Pull request by alan-geller**
Adds some unit tests for partial applications and %Callable values

### CODE REVIEW

File: src/QsCompiler/QirGenerator/GenerationContext.cs [hunk 1]
```diff
@@ -515,40 +515,71 @@ IEnumerable<ITypeRef> ArgTupleToTypes(QsArgumentTuple arg)
...
+                            Value asStructPointer = this.CurrentBuilder.BitCast(val,
+                                tupleTypeRef.CreatePointerType());
+                            var indices = new Value[] { this.CurrentContext.CreateConstant(0L),
+                                                    this.CurrentContext.CreateConstant(1) };
+                            for (var i = 0; i < n; i++)
+                            {
+                                indices[1] = this.CurrentContext.CreateConstant(i + 1);
+                                Value ptr = this.CurrentBuilder.GetElementPtr(tupleTypeRef, asStructPointer, indices);
+#pragma warning disable CS0618 // Type or member is obsolete -- computing the type that ptr points to is tricky, so we just rely on it's .NativeType
+                                var subVal = this.CurrentBuilder.Load(ptr);
```
Comment (bettinaheim) [inline src/QsCompiler/QirGenerator/GenerationContext.cs:538]: Is Load obsolete? Why are we using something obsolete here?

File: src/QsCompiler/Tests.Compiler/QirTests.fs [hunk 2]
```diff
@@ -136,10 +160,18 @@ let ``QIR bools`` () =
 let ``QIR bigints`` () =
     qirTest false "TestBigInts"
 
+[<Fact>]
+let ``QIR controlled partial applications`` () =
```
Comment (bettinaheim) [inline src/QsCompiler/Tests.Compiler/QirTests.fs:164]: Just for the sake of a decent test coverage it would be nice to also have a test with some adjoint and controlled adjoint applications.

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+// Licensed under the MIT License.
+
+namespace Microsoft.Quantum.Testing.QIR
+{
+    open Microsoft.Quantum.Intrinsic;
+
+    @EntryPoint()
+    operation TestPartials () : Bool
+    {
+        let rotate = Rz(0.25, _);
```
Comment (bettinaheim) [inline src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs:11]: It would be nice to add a couple more scenarios here; e.g. for an operation that usually take a nested tuple Op(a, (b, c)) I'd test the cases Op(_, (b, c)), Op(a, (_, c)), Op(a, (_,_)), and Op (a, _).

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        let unrotate = Adjoint rotate;
+
+        for (i in 0..100)
+        {
+            using (qb = Qubit())
+            {
+                rotate(qb);
+                unrotate(qb);
```
Comment (bettinaheim) [inline src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestPartials.qs:19]: It would also be good to check that partial application after functor application works as expected, in particular, I'd check that both Controlled Op(_, (a, (b, c))) and Controlled Op(qs, (a, (_, c))) work as expected.

File: src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestControlled.qs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        let k2 = K(_, 2);
+        let ck2 = Controlled k2;
+        let ck1 = K(_, _);
+
+        for (i in 0..100)
+        {
+            using ((ctrls, qb) = (Qubit[2], Qubit()))
+            {
+                ck2(ctrls, qb);
```
Comment (bettinaheim) [inline src/QsCompiler/Tests.Compiler/TestCases/QirTests/TestControlled.qs:55]: An edge case that I think is still worth covering for the sake of a good test coverage is invoking the controlled specialization for an operation that takes a Unit argument.

**Comment by bettinaheim** [comment]
Incorporated into https://github.com/microsoft/qsharp-compiler/pull/764.
